elastic-tube-3d/solid-fenics/solid.py

```diff
@@ -0,0 +1,229 @@
+# Import required libs
+from fenics import Constant, Function, AutoSubDomain, VectorFunctionSpace, interpolate, \
+    TrialFunction, TestFunction, Point, Expression, DirichletBC, nabla_grad, \
+    Identity, inner, dx, ds, sym, grad, lhs, rhs, dot, File, solve, assemble_system
+from mshr import Cylinder, generate_mesh
+from ufl import nabla_div
+import numpy as np
+from fenicsprecice import Adapter
+import math
+
+
+# define the two kinds of boundary: clamped and coupling Neumann Boundary
+def clamped_boundary(x, on_boundary):
+    """
+    Filter nodes at both ends of tube as they are fixed
+    """
+    tol = 1E-14
+    return on_boundary and ((x[2] - 0.0) < tol) and ((x[2] - L) < tol)
```

Comment:
```suggestion
    return on_boundary and ((x[2] - 0.0) < tol) or ((x[2] - L) < tol)
```

I think this might be the error. You are using multiple `and`s here. This means only points that are at both ends of the tube are fixed.

Reply:
To make it less errorprone, you could also do something like:

```
is_inflow = (x[2] - 0.0) < tol
is_outflow = (x[2] - L) < tol
return on_boundary and (is_inflow or is_outflow)
```